diesel_cli/src/migrations/mod.rs

```diff
@@ -273,55 +273,75 @@ fn redo_migrations<Conn, DB>(
+    };
+    let should_use_not_use_transaction = versions_to_revert.iter().any(|v| {
+        migrations
+            .get(v)
+            .map(|m| !m.metadata().run_in_transaction())
+            .unwrap_or_default()
+    });
+
+    let migrations_inner =
+        |harness: &mut HarnessWithOutput<Conn, _>|
+         -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+            let reverted_versions = if args.get_flag("REDO_ALL") {
+                harness.revert_all_migrations(migrations_dir.clone())?
+            } else {
+                let number = args.get_one::<u64>("REDO_NUMBER").unwrap();
+
+                (0..*number)
+                    .filter_map(|_| {
+                        match harness.revert_last_migration(migrations_dir.clone()) {
+                            Ok(v) => Some(Ok(v)),
+                            Err(e) if e.is::<MigrationError>() => {
+                                match e.downcast_ref::<MigrationError>() {
+                                    // If n is larger then the actual number of migrations,
+                                    // just stop reverting them
+                                    Some(MigrationError::NoMigrationRun) => None,
+                                    _ => Some(Err(e)),
+                                }
                             }
+                            Err(e) => Some(Err(e)),
                         }
-                        Err(e) => {
-                            Some(Err(e))
-                        }
-                    }
-                })
-                .collect::<Result<Vec<_>, _>>()?
-        };
+                    })
+                    .collect::<Result<Vec<_>, _>>()?
+            };
 
-        let mut migrations = MigrationSource::<DB>::migrations(&migrations_dir)?
-            .into_iter()
-            .map(|m| (m.name().version().as_owned(), m))
-            .collect::<HashMap<_, _>>();
+             let mut migrations = MigrationSource::<DB>::migrations(&migrations_dir)
+                 .unwrap_or_else(handle_error)
+                 .into_iter()
+                 .map(|m| (m.name().version().as_owned(), m))
+                 .collect::<HashMap<_, _>>();
 
-        let mut migrations = reverted_versions
-            .into_iter()
-            .map(|v| {
-                migrations
-                    .remove(&v)
-                    .ok_or_else(|| MigrationError::UnknownMigrationVersion(v.as_owned()))
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-        // Sort the migrations by version to apply them in order.
-        migrations.sort_by_key(|m| m.name().version().as_owned());
+            let mut migrations = reverted_versions
+                .into_iter()
+                .map(|v| {
+                    migrations
+                        .remove(&v)
+                        .ok_or_else(|| MigrationError::UnknownMigrationVersion(v.as_owned()))
+                })
+                .collect::<Result<Vec<_>, _>>()?;
```

Comment:
can we create a new function for this. something like, `remove_reverted_migrations(migrations)`

Reply:
As far as I'm concerned I typically don't like functions that are only called once, as you can achieve the same clarity by putting a comment just above the section, whereas having an extra function:
- complexifies understanding precisely what it does and making sure everything is correct (review) because you have to always keep track mentally of where you are in the call tree as you go to understand in practice exactly what will be linearly executed, instead of just scrolling up and down
- complexifies modifying the code because whenever you consider updating a node (function) you have to check for each of the calling nodes up to your entry point if they aren't called by anyone else to be sure of what else you may be affecting. When there's a single function, it's trivial to identify that that is not the case, and that makes edits easier

Reply:
The comments are helpful, thanks.